src/mlpack/methods/reinforcement_learning/environment/frozen_lake.hpp

```diff
@@ -0,0 +1,470 @@
+  }
+
+  /**
+   * This function checks if the agent has reached the terminal state.
+   *
+   * @param state The desired state.
+   * @return true if state is a terminal state, otherwise false.
+   */
+  bool IsTerminal(const State& state) const
+  { 
+    if (maxSteps != 0 && stepsPerformed >= maxSteps)
+    {
+      Log::Info << "Episode terminated due to the maximum number of steps"
+          "being taken.\n";
+      return true;
+    }
+    else if (boardDescription(state.CurRow(), state.CurCol()) == tiles::Hole)
+    {
+      Log::Info << "Episode terminated due to agent falling in the hole.\n";
+      return true;
+    }
+    else if (boardDescription(state.CurRow(), state.CurCol()) == tiles::Goal)
+    {
+      Log::Info << "Episode terminated, agent reached the goal.\n";
+      return true;
+    }
+    return false;
+  }
+
+  //! Get the number of steps performed.
+  size_t StepsPerformed() const { return stepsPerformed; }
+
+  //! Get the maximum number of steps allowed.
+  size_t MaxSteps() const { return maxSteps; }
+
+  //! Set the maximum number of steps allowed.
+  size_t& MaxSteps() { return maxSteps; }
+
+  //! Get the board description.
+  // std::vector<std::vector<char>> Description() const {return boardDescription;}
+  arma::Mat<size_t> Description() const {return boardDescription;}
+
+  //! Set the board description.
+  // std::vector<std::vector<char>>& Description() {return boardDescription;}
+  arma::Mat<size_t>& Description() {return boardDescription;}
+
+  enum tiles
+  {
+    Start, Goal, Frozen, Hole
+  };
+
+ private:
+  /**
+   * Utility function that helps generate random environment board.
+   * 
+   * @return board, a 2D array of characters that hold the board description.
+   *    'S' is the Start Tile, the starting position of the agent. 0
+   *    'G' is the Goal Tile, the agent can walk into a goal tile to win. 1
+   *    'F' is a Frozen Tile, the agent can walk on it. 2
+   *    'H' is a Hole Tile, the agent can fall into it and the game ends. 3
+   */
+  arma::Mat<size_t> generateRandomBoard()
+  {
+    bool valid = false;
+    // std::vector<std::vector<char>> Board;
+    arma::Mat<size_t> Board(height, width);
+    //! TODO: Tentative, implement max step to control whether
+    // we can never generate a possible board.
+    while (!valid)
+    {
+      // Randomly choose tiles in the board (discrete random distribution).
+      Board = genBoardHelper(height, width, platformRate);
+      valid = dfsHelper(Board, height, width);
+    }
+    return Board;
+  }
+
+  /**
+   * Perform depth-firsth search to see if the environment board has 
+   * a solution or not.
+   * 
+   * @param candidateBoard Environment board description.
+   * @param height Height of the environment board.
+   * @param width Width of the environment board.
+   * @return true if there is a solution, false otherwise. 
+   */
+  static bool dfsHelper(arma::Mat<size_t> candidateBoard, size_t height, size_t width)
```

Comment:
```suggestion
  static bool DfsHelper(arma::Mat<size_t> candidateBoard, size_t height, size_t width)
```

We use upper camel-casing in mlpack.

Reply:
sure! Will change!